### ISSUE TITLE
Model size is getting bigger after training

### ISSUE THREAD

**Issue by Shahrullo** [ultralytics/ultralytics#1736]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

In some projects, the model size is getting bigger. For example, I used the pre-trained model `YOLOv8x` which is **136MB** for custom data training. After training, now I have model **546MB**. Is this a known bug or just the nature of YOLOv8 models?

**Comment by HornGate** [comment]
> ### Search before asking
> 
>     * [x]  I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.
> 
> 
> ### Question
> 
> In some projects, the model size is getting bigger. For example, I used the pre-trained model `YOLOv8x` which is **136MB** for custom data training. After training, now I have model **546MB**. Is this a known bug or just the nature of YOLOv8 models?
> ### Additional
> 
> _No response_

When you finished training, the final model will be 136 MB.

**Comment by glenn-jocher** [comment]
You're welcome, @Shahrullo! Happy to help. If you have any more questions or issues, please don't hesitate to reach out. Have a great day!

**Comment by Shahrullo** [comment]
Thank you @glenn-jocher. I will. Have a great day too and good luck with your progress. Happy to see that YOLOv8 is applicable for Segmentation and Pose Estimation

**Comment by glenn-jocher** [comment]
Thank you for your kind words, @Shahrullo. Yes, YOLOv8 can be applied beyond object detection tasks, such as segmentation and pose estimation. This is possible thanks to the versatility and flexibility of the underlying architecture, which can be adapted to different types of models and tasks. However, it is worth noting that training these models might require a different setup than object detection models, and may require custom modifications and training procedures. It is important to consult the relevant documentation and other resources when training these kinds of models.

**Comment by Shahrullo** [comment]
Thank you @glenn-jocher. Of course, I will refer to the documentation~~

**Comment by glenn-jocher** [comment]
Hello @Shahrullo,

Thank you for your question. To answer your question, yes, YOLOv8 is completely open-source and maintained under the GPL-3.0 license, so you are free to use it for any purpose, including commercial applications, as long as you comply with the GPL-3.0 license requirements. 

Furthermore, we always recommend that users provide attribution to the YOLOv8 project and its contributors verbally or in their documentation, as this helps us continue to improve and support the project in the future. Does that answer your question? Please let me know if you have any further questions or concerns.

**Comment by glenn-jocher** [comment]
@Shahrullo ah I just realized, your checkpoints are larger because they have not completed training. Here is a guide from YOLOv5:

👋 Hello! Thanks for asking about **training checkpoints**.  YOLOv5 🚀 checkpoints should be about 4X the size of final trained checkpoints, as they carry not just a FP16 model, but a FP16 [EMA](https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage) and an FP32 optimizer of the same size as the model (each model parameter has its own FP32 gradient saved within the optimizer). Final checkpoints contain only an FP16 model, with the EMA and optimizer both stripped after the final epoch of training:
https://github.com/ultralytics/yolov5/blob/b4a29b5a8d63a8c2d4a8929942b44e8969c5dddd/train.py#L423-L425

The `strip_optimizer()` function updates the checkpoint dictionary value for `model`, replacing it with `ema`, and sets `ema` and `optimizer` keys to `None`, which will reduce the checkpoint size by 3/4.
https://github.com/ultralytics/yolov5/blob/b4a29b5a8d63a8c2d4a8929942b44e8969c5dddd/utils/general.py#L741-L755

Final trained checkpoints after `strip_optimizer()` should match the README table sizes below.
<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

You can also run `strip_optimizer()` manually on any checkpoint to convert it into finalized weights ready to train a new model:
```python
from utils.general import strip_optimizer

strip_optimizer('path/to/best.pt')
!python train.py --weights path/to/best.pt  # use best.pt to train a new model
```

### Model Exports

Exported models may be larger or smaller than their PyTorch starting points depending on their quantization. i.e. FP16 models will be the same size, FP32 models will be 2X the size, and INT8 models will be 0.5X the size. See YOLOv5 [Export tutorial](https://github.com/ultralytics/yolov5/issues/251) for quantization options on model export.

Good luck 🍀 and let us know if you have any other questions!

**Comment by Shahrullo** [comment]
Hi @glenn-jocher, thanks a lot for letting me know about it:) Now it is more clear to me!

**Comment by glenn-jocher** [comment]
Hello @Shahrullo, I'm glad to hear that I was able to help you! It's not uncommon for training checkpoints to be larger than the final trained checkpoints, and in the case of YOLOv5, the final trained checkpoints should be about 1/4 of the size of the training checkpoints, once the optimizer and EMA have been stripped out. This is because the checkpoints carry not only a model, but also an optimizer and EMA to facilitate training. 

It's important to keep in mind that the final size of the exported model will depend on the quantization used on export. For example, an INT8 model will be smaller than an FP32 model, while an FP16 model will have the same size as the original PyTorch model. If you have any more questions or concerns, feel free to ask!
